src/com/owncloud/android/providers/DocumentsStorageProvider.java

```diff
@@ -0,0 +1,208 @@
+/**
+ *   ownCloud Android client application
+ *
+ *   @author Bartosz Przybylski
+ *   Copyright (C) 2015  Bartosz Przybylski
+ *   Copyright (C) 2015 ownCloud Inc.
+ *
+ *   This program is free software: you can redistribute it and/or modify
+ *   it under the terms of the GNU General Public License version 2,
+ *   as published by the Free Software Foundation.
+ *
+ *   This program is distributed in the hope that it will be useful,
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *   GNU General Public License for more details.
+ *
+ *   You should have received a copy of the GNU General Public License
+ *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package com.owncloud.android.providers;
+
+import android.accounts.Account;
+import android.annotation.TargetApi;
+import android.content.ContentResolver;
+import android.content.Intent;
+import android.content.res.AssetFileDescriptor;
+import android.database.Cursor;
+import android.graphics.Point;
+import android.os.Build;
+import android.os.CancellationSignal;
+import android.os.ParcelFileDescriptor;
+import android.provider.DocumentsProvider;
+
+import com.owncloud.android.authentication.AccountUtils;
+import com.owncloud.android.datamodel.FileDataStorageManager;
+import com.owncloud.android.datamodel.OCFile;
+import com.owncloud.android.files.services.FileDownloader;
+import com.owncloud.android.providers.cursors.FileCursor;
+import com.owncloud.android.providers.cursors.RootCursor;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Vector;
+
+@TargetApi(Build.VERSION_CODES.KITKAT)
+public class DocumentsStorageProvider extends DocumentsProvider {
+
+    private FileDataStorageManager mCurrentStorageManager = null;
+    private static Map<Long, FileDataStorageManager> mRootIdToStorageManager;
+
+    @Override
+    public Cursor queryRoots(String[] projection) throws FileNotFoundException {
+        initiateStorageMap();
+
+        final RootCursor result = new RootCursor(projection);
+
+        for (Account account : AccountUtils.getAccounts(getContext()))
+            result.addRoot(account, getContext());
+
+        return result;
+    }
+
+    @Override
+    public Cursor queryDocument(String documentId, String[] projection) throws FileNotFoundException {
+        final long docId = Long.parseLong(documentId);
+        updateCurrentStorageManagerIfNeeded(docId);
+
+        final FileCursor result = new FileCursor(projection);
+        result.addFile(mCurrentStorageManager.getFileById(docId));
```

Comment:
Here it is been generated sometimes a NullPointer exception, because `mCurrentStorageManager` is found in the StorageManager Map with a different id of the current storage managers.